src/ir/auth_logic/universe_relation_insertion.cc

```diff
@@ -0,0 +1,249 @@
+    }
+  }
+
+  std::vector<SaysAssertion> universe_defining_says_assertions;
+  for (const auto& prin : constant_principals) {
+    for (const auto& typing : type_env.literal_type_map()) {
+      universe_defining_says_assertions.push_back(SaysAssertion(
+          prin,  // speaker
+          {Assertion(Fact(
+              {}, MakeUniverseMembershipFact(typing.first, typing.second)))}));
+    }
+  }
+  return universe_defining_says_assertions;
+}
+
+// This visitor is used only to implement GetUniverseDeclarations
+class UniverseDeclarationGenerationVisitor
+    : public AuthLogicAstTraversingVisitor<
+          UniverseDeclarationGenerationVisitor,
+          absl::flat_hash_set<datalog::RelationDeclaration>> {
+ private:
+  using DeclSet = absl::flat_hash_set<datalog::RelationDeclaration>;
+  DeclSet GetDefaultValue() override { return {}; }
+
+  DeclSet CombineResult(DeclSet left, DeclSet right) override {
+    left.merge(std::move(right));
+    return left;
+  }
+
+  DeclSet Visit(
+      const datalog::RelationDeclaration& relation_declaration) override {
+    DeclSet universe_declarations;
+    for (const auto& argument : relation_declaration.arguments()) {
+      universe_declarations.insert(
+          TypeToUniverseDeclaration(argument.argument_type()));
+    }
+    return universe_declarations;
+  }
+};
+
+static absl::flat_hash_set<datalog::RelationDeclaration>
+GetUniverseDeclarations(const Program& prog) {
+  UniverseDeclarationGenerationVisitor gen;
+  absl::flat_hash_set<datalog::RelationDeclaration> ret = prog.Accept(gen);
+  // It is possible that universes need to be created
+  // for numbers and principals even if they are not
+  // referenced in relation declarations in the front-end
+  absl::flat_hash_set<datalog::RelationDeclaration> builtin_universes = {
+      TypeToUniverseDeclaration(datalog::ArgumentType(
+          datalog::ArgumentType::Kind::kNumber, "Number")),
+      TypeToUniverseDeclaration(datalog::ArgumentType(
+          datalog::ArgumentType::Kind::kPrincipal, "Principal")),
+  };
+  ret.merge(builtin_universes);
+  return ret;
+}
+
+//----------------------------------------------------------------------------
+// Grounding Condition Transformation
+//----------------------------------------------------------------------------
+// This class transforms SaysAssertions so that if
+// there are any potentially ungrounded variables on the LHS of an assertion,
+// conditions are added to the RHS that check that this variable
+// is in a universe. As a way of safely overapproximating the ungrounded
+// variables, this just assumes any *non-constant* arguments in the RHS
+// are ungrounded, and adds conditions for them to the LHS.
+// It implements just one public method,
+// `AddUniverseConditions(SaysAssertion says_assertion)`
+// that returns a transformed version of the says assertion.
+// Internally, it extends a visitor for a part of this translation
+class UniverseConditionTransformer
```

Comment:
nit: a more appropriate name? e.g., `SaysAssertionTransformer`

Reply:
OK I called it `GroundingConditionTransformer`. ?